rfcs/0004-policy-sync-v1.md

```diff
@@ -0,0 +1,173 @@
+# RFC - Policy Sync
+
+- Feature Name: `policy_sync_v1`
+- Start Date: 2023-10-10
+- RFC PR: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/pull/0000)
+- Issue tracking: [https://github.com/Kuadrant/architecture/issues/26](https://github.com/Kuadrant/architecture/issues/26)
+
+# Summary
+[summary]: #summary
+
+The ability for the Multicluster Gateway Controller to sync policies defined in
+the hub cluster downstream to the spoke clusters, therefore allowing all policies
+to be defined in the same place. These policies will be reconciled by the downstream
+Gateway controller.
+
+# Nomenclature
+
+* Policy: When refering to a Policy, this document is refering to a Gateway API
+  policy as defined in the Policy Attachment Model. The Multicluster Gateway Controller
+  relies on [OCM]() as a Multicluster solution, which defines its own unrelated
+  set of Policies and Policy Framework. Unless explicitely mentioned, this document
+  refers to Policies as Gateway API Policies.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Kuadrant's support for the Policy Attachment Model can be divided in
+two categories:
+* Policies targeting the Multicluster Gateway, defined in the hub cluster and
+  reconciled by the Multicluster Gateway Controller
+* Policies targeting the downstream Gateway, defined in the spoke clusters and
+  reconciled by the downstream Gateway controllers.
+
+In a realistic multicluster scenario where multiple spoke clusters are present, the management of these policies can become tedious and error-prone, as policies have
+to be defined in the hub cluster, as well as replicated in the multiple spoke clusters.
+
+As Kuadrant users:
+* Gateway-admin has a set of homogeneous clusters and needs to apply per cluster rate limits across the entire set.
```

Comment:
From the point of view of a user who wants to create and manage policies for their Gateway, the clusters are homogeneous. The underlying infrastructure is irrelevant for that purpose.

As for the case of applying specific policies to a subset of the clusters, that's why the hierarchy and overriding section is for. A user can create a RateLimitPolicy in the hub cluster to apply in all clusters, and create a RateLimitPolicy in only one of the clusters that overrides the synced Policy.

Basically, the purpose of this RFC is to allow to create policies that target all spoke clusters, in a centralised way, with the possibility of them being overriden

Reply:
At some point the capacity of the underlying infrastructure becomes relevant, and I believe this point is at the level of the gateway or gateway class. What the actual underlying infrastructure is (AWS, GCP), that is irrelevant for this purpose.

As of today the RateLimitPolicy and AuthPolicy don't support defaults and overrides, which I am assuming would be required for this. There is talks of getting that support there, so I am ok with this RFC being a thing before that is there. 

Where I do see a problem with the use case of using defaults and overrides, the Kuadrant Policies can only target one resource. And I have not heard any talks about changing this. With the example of creating a RateLimitPolicy on the hub targeting the gateway and is copied to the spokes. If the Gateway owner wanted to override that policy on a per spoke bases, the overrides would need to be added to the RateLimitingPolicy attached to the HTTPRoute. Which maybe owned by different orgs. 

To create the overrides in a RateLimitPolicy on the Gateway, it would require a RateLimitPolicy targeting the GatewayClass which is not supported today.

Reply:
I think there's a mixup about the concept of overrides. When I talk about overrides in this RFC I mean adding support in the Kuadrant operator to ignore a policy that has been synced from the hub if it conflicts with a policy created in the spoke. That support would be part of the implementation of this RFC

Maybe I could add to the document an item to the nomenclature section to avoid confusion

Reply:
> Where I do see a problem with the use case of using defaults and overrides, the Kuadrant Policies can only target one resource. And I have not heard any talks about changing this. With the example of creating a RateLimitPolicy on the hub targeting the gateway and is copied to the spokes. If the Gateway owner wanted to override that policy on a per spoke bases, the overrides would need to be added to the RateLimitingPolicy attached to the HTTPRoute. Which maybe owned by different orgs.

If they wanted to do something per cluster, they would go back to creating the policy on a per cluster basis. This is what they have to do now. 
If they want something to be replicated accross many clusters with the same definition but define that targeting the multicluster gateway, they would define the policy in the hub targeting the multicluster gateway

Reply:
Some of this really is also to do with the fact we do not let the user set a counter strategy in anyway. 
IE ```counterStrategy``` for limit A is clusterUnique for limit B is shared. being able to define the counter strategy would allow you decide if a somethign should be shared or unique
Also it is possible in RLP to set a limit that only get triggered "when" a certain value is passed. So if you had a unique value being passed from each gateway you could use that to distinguish limits (IE if gateway A set limit to x every y minutes)
Finally another option here is for the policy sync to add some form of label to identify the cluster it came from, these label could be injected into the config for the WASMPlugin and used in conditions.

All of this is to say, that the syncing of the policy is not necessarilly where we want to or where we need to solve changing a set of limits for a particular cluster

Reply:
https://github.com/Kuadrant/kuadrant-operator/issues/271